### DISCUSSION TITLE
Unique Id

### DISCUSSION THREAD

**Discussion by 47vigen** [facebook/lexical#2665] [Q&A] [answered]
Im very looking for how can i define unique id for each nodes in lexical same as this link for tiptap... 🙂

https://tiptap.dev/api/extensions/unique-id

**Comment by thegreatercurve** [comment, marked as answer]
It's not customizable, but we already create and store a unique `key` on each node instance. You should be able to access the unique key/ID with `node.getKey()`.

**Reply by jiangyuhang0506** [reply]
Will this ID not change, for example, if I modify it, is the key for each node fixed?

**Comment by 47vigen** [comment]
> It's not customizable, but we already create and store a unique `key` on each node instance. You should be able to access the unique key/ID with `node.getKey()`.

We want this to communicate with our backend, the question that remains is to parse a json with the same keys. Do you have any opinion about this?

**Reply by thegreatercurve** [reply]
>  We want this to communicate with our backend

I think I'm struggling to understand the use case, but having unique IDs shouldn't make much difference if we're talking about maintaining consistency between the front and backends.

It's up to the developer how they want to [serialize/deserialize editor state to JSON](https://lexical.dev/docs/concepts/serialization), and even though we don't include the keys in the default nodes on serialization, you can use the generated HTML or JSON to check equality between different editor states on the front or backend.

**Reply by khanakia** [reply]
@thegreatercurve I want to get the unique id while using export To JSON and then again i want to lexical to use the same unique id when importing from JSON.

I do not see a way to make it possible as lexical automatically handles the id generation

As you can see in below i want something like this `id` so i do want to keep the id same always for that node while import and export from JSON
```
const editorState = {
  "root": {
    "children": [
      {
        "children": [
          {
            "id": "5y2edpsiz5k",
            "detail": 0,
            "format": 0,
            "mode": "normal",
            "style": "",
            "text": "To generate a JSON snapshot from an editor.",
            "type": "text",
            "version": 1
          },
        ],
        "direction": "ltr",
        "format": "",
        "indent": 0,
        "type": "paragraph",
        "version": 1
      }
    ],
    "direction": "ltr",
    "format": "",
    "indent": 0,
    "type": "root",
    "version": 1
  }
}
```

**Reply by MaxMernikov** [reply]
@khanakia Could you please share your solution if you managed to do importing from JSON with unique id and back into JSON

**Reply by khanakia** [reply]
You can check the attached file for reference code
[blockquote.zip](https://github.com/facebook/lexical/files/13632127/blockquote.zip)

**Comment by rohamgoudarz** [comment]
Is there any documentation about best practices to save Lexical content to DB and retrieve it?
Because as far as I can see, there aren't any unique-id (Remirror approach) to mark each children.

**Reply by thegreatercurve** [reply]
You define how you want to serialize each node on the node class itself. Your can then serialize it to JSON or HTML: https://lexical.dev/docs/concepts/serialization

**Comment by jiangyuhang0506** [comment]
How to add a unique ID to JSON, or how to make logical judgments before and after modifying JSON
